### PR TITLE
Added dynamic forward proxy cluster

### PR DESCRIPTION
**Pull request by lukidzi**

### CODE REVIEW

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt [hunk 1]
```diff
@@ -425,4 +437,54 @@ class EnvoyClustersFactory(
                     )
             )
     }
+
+    private fun createDynamicForwardProxyCluster(): Cluster {
+        return Cluster.newBuilder()
+            .setName(properties.dynamicForwardProxy.clusterName)
+            .setConnectTimeout(Duration.newBuilder().setSeconds(1))
```
Comment (slonka) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/clusters/EnvoyClustersFactory.kt:444]: This should be configureable.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/EnvoyListenersFactory.kt [hunk 2]
```diff
@@ -70,29 +71,36 @@ class EnvoyListenersFactory(
...
-                    .addTlsCertificateSdsSecretConfigs(SdsSecretConfig.newBuilder()
-                            .setName(tlsProperties.tlsCertificateSecretName)
-                            .build()
-                    )
-                    .setValidationContextSdsSecretConfig(SdsSecretConfig.newBuilder()
-                            .setName(tlsProperties.validationContextSecretName)
-                            .build()
-                    )
-            )
+        .setRequireClientCertificate(requireClientCertificate)
```
Comment (slonka) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/EnvoyListenersFactory.kt:78]: This is just formatting right?

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        val envoy = EnvoyExtension(envoyControl, config = AdsDynamicForwardProxy)
+    }
+
+    @Test
+    fun `should allow to request domains with suffix com`() {
+        // given
+        untilAsserted {
+            // when
+            val reachableDomainResponse = envoy.egressOperations.callDomain("www.example.com")
+            val reachableDomainResponse2 = envoy.egressOperations.callDomain("www.wp.pl")
```
Comment (slonka) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt:46]: Please do not use specific domains other than `example.com` `example.pl`

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        @RegisterExtension
+        val envoy = EnvoyExtension(envoyControl, config = AdsDynamicForwardProxy)
+    }
+
+    @Test
+    fun `should allow to request domains with suffix com`() {
+        // given
+        untilAsserted {
+            // when
+            val reachableDomainResponse = envoy.egressOperations.callDomain("www.example.com")
```
Comment (slonka) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt:45]: If I'm not missing anything this actually calls `www.example.com` right? If so please change it. I would add example.com to `hosts` file (there are already helper functions for that), and have an http server on that IP that respons.

**Review by slonka — CHANGES_REQUESTED** [review]
There are a couple of small things. Also why do we need the pattern (can't we just allow all domains)? Envoy requirers it? Another thing: do we have a way to disable adding dynamic forward proxy as a spring property?

File: docs/configuration.md [hunk 5]
```diff
@@ -74,7 +74,11 @@ Property
...
-     
+**envoy-control.envoy.snapshot.cluster-name**                                                                | Dynamic forward proxy cluster name                                                                                                                                                                                    | dynamic_forward_proxy_cluster
+**envoy-control.envoy.snapshot.dns-lookup-family**                                                           | DNS lookup address family                                                                                                                                                                                             | V4_ONLY
+**envoy-control.envoy.snapshot.max-cached-hosts**                                                            | The maximum number of hosts that the cache will hold                                                                                                                                                                  | 1024
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply docs/configuration.md:79]: We don't know exact number. https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/http/http_proxy#memory-usage-details

**Comment by lukidzi** [comment]
> There are a couple of small things. Also why do we need the pattern (can't we just allow all domains)? Envoy requirers it? Another thing: do we have a way to disable adding dynamic forward proxy as a spring property?

We don't want to use `*` because it would try to do requests to services which we don't support. We have a default_route which handle unknown services and return 404. We don't need to disable dynamic forward proxy, because you can only get it when specific property is passed from Envoy.

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt [hunk 6]
```diff
@@ -35,19 +40,27 @@ class DynamicForwardProxyTest {
...
+        fun setup() {
+            envoy.container.addDnsEntry("my.example.com", httpsService.container().ipAddress())
+        }
     }
 
     @Test
     fun `should allow to request domains with suffix com`() {
         // given
-        untilAsserted {
+        untilAsserted(wait = Duration.ofSeconds(1200)) {
```
Comment (slonka) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt:58]: Why does it take so long?

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt [hunk 7]
```diff
@@ -31,6 +31,7 @@ class EnvoyContainer(
         private const val EXTRA_DIR = "envoy/extra"
         private const val EXTRA_DIR_DEST = "/etc/envoy/extra"
 
+        const val ENVOY_UID_ENV_NAME = "ENVOY_UID"
```
Comment (slonka) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt:34]: what does this do?

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt [hunk 8]
```diff
@@ -69,6 +70,11 @@ class EnvoyContainer(
         )
     }
 
+    fun addDnsEntry(host: String, ip: String) {
```
Comment (slonka) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt:73]: What about cleanup? Maybe we should clean up those entries per test class/method. WDYT?

**Review by slonka — COMMENTED** [review]
Looks alright, just a couple of questions.

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt (hunk 6 above)
Comment (lukidzi) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/DynamicForwardProxyTest.kt:58]: My bad, was testing setup and forgot to change

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt (hunk 8 above)
Comment (lukidzi) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt:73]: added

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt (hunk 7 above)
Comment (lukidzi) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/envoy/EnvoyContainer.kt:34]: https://www.envoyproxy.io/docs/envoy/latest/start/docker#permissions-for-running-docker-envoy-containers-as-a-non-root-user

**Review by slonka — DISMISSED** [review]
LGTM

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt [hunk 9]
```diff
@@ -125,6 +141,17 @@ private fun validateDomainFormat(
     }
 }
 
+private fun validateDomainPatternFormat(
+    it: RawDependency
+) {
+    val domainPattern = it.domainPattern.orEmpty()
+    if (domainPattern.startsWith("http://") || domainPattern.startsWith("https://")) {
+        throw NodeMetadataValidationException(
+            "Unsupported format for domainPattern"
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadata.kt:150]: This is an unspecific error message to a very specific error.
```suggestion
            "Unsupported format for domainPattern: domainPattern cannot contain a schema like http:// or https://"
```

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/routes/EnvoyEgressRoutesFactoryTest.kt [hunk 10]
```diff
@@ -20,7 +20,7 @@ internal class EnvoyEgressRoutesFactoryTest {
     val clusters = listOf(
         RouteSpecification(
             clusterName = "srv1",
-            routeDomain = "srv1",
+            routeDomains = listOf("srv1"),
```
Comment (pzmi) [inline envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/routes/EnvoyEgressRoutesFactoryTest.kt:23]: This hadn't broken any tests? Interesting.

**Review by pzmi — APPROVED** [review]
LGTM. Just a small comment.

**Review by pzmi — APPROVED** [review]

**Review by slonka — APPROVED** [review]

**Pull request merged**